arviz/kde_utils.py

```diff
@@ -0,0 +1,812 @@
+    [1] C.C Taylor (2008). Automatic bandwidth selection
+    for circular density estimation.
+    Computational Statistics and Data Analysis, 52, 7, 3493–3500.
+    """
+    x_len = len(x)
+    kappa = kappa_mle(x)
+    num = 3 * x_len * kappa ** 2 * ive(2, 2 * kappa)
+    den = 4 * np.pi ** 0.5 * ive(0, kappa) ** 2
+    return (num / den) ** 0.4
+
+
+BW_METHODS = {
+        "scott": bw_scott,
+        "silverman": bw_silverman,
+        "isj": bw_isj,
+        "experimental": bw_experimental
+    }
+
+
+def select_bw_method(method):
+    """
+    Selects a function to compute the bandwidth.
+    Also checks method `bw` is correctly specified.
+    Otherwise, throws an error.
+
+    Parameters
+    ----------
+    method : str
+        Method to estimate the bandwidth.
+
+    Returns
+    -------
+    bw_fun: function
+        Function to compute the bandwidth.
+    """
+
+    method_lower = method.lower()
+    if method_lower not in BW_METHODS.keys():
+        raise ValueError((
+            f"Unrecognized bandwidth method.\n"
+            f"Input is: {method}.\n"
+            f"Expected one of: {list(BW_METHODS.keys())}."
+        ))
+    bw_fun = BW_METHODS[method_lower]
+    return bw_fun
+
+
+def get_bw(x, bw, grid_counts=None, x_std=None, x_range=None):
+    """
+    Computes bandwidth for a given data `x` and `bw`.
+    Also checks `bw` is correctly specified.
+
+    Parameters
+    ----------
+    x : 1-D numpy array
+        1 dimensional array of sample data from the
+        variable for which a density estimate is desired.
+    bw: int, float or str
+        If numeric, indicates the bandwidth and must be positive.
+        If str, indicates the method to estimate the bandwidth.
+
+    Returns
+    -------
+    bw: float
+        Bandwidth
+    """
+    if isinstance(bw, bool):
+        raise ValueError((
+            f"`bw` must not be of type `bool`.\n"
+            f"Expected a positive numeric or one of the following strings:\n"
+            f"{list(BW_METHODS.keys())}."))
+    if isinstance(bw, (int, float)):
+        if bw < 0:
+            raise ValueError(f"Numeric `bw` must be positive.\nInput: {bw:.4f}.")
+    elif isinstance(bw, str):
+        bw_fun = select_bw_method(bw)
+        bw_lower = bw.lower()
+        if bw_lower == "isj":
+            bw = bw_fun(x, grid_counts, x_range)
+        elif bw_lower in ["scott", "silverman"]:
+            bw = bw_fun(x, x_std)
+        elif bw_lower == "experimental":
+            bw = bw_fun(x, grid_counts, x_std, x_range)
+    else:
+        raise ValueError((
+            f"Unrecognized `bw` argument.\n"
+            f"Expected a positive numeric or one of the following strings:\n"
+            f"{list(BW_METHODS.keys())}."))
+    return bw
+
+
+# Misc utils
+def vonmises_pdf(x, mu, kappa):
+    assert kappa > 0, "Argument 'kappa' must be positive."
+    pdf = 1 / (2 * np.pi * ive(0, kappa)) * np.exp(np.cos(x - mu) - 1) ** kappa
+    return pdf
+
+
+def circular_mean(x):
```

Comment:
Seems like custom function is faster (it was not my main goal when I implemented it, I just didn't know scipy had its own function)
![Screenshot from 2020-07-17 00-31-40](https://user-images.githubusercontent.com/25507629/87745719-11d6dc80-c7c5-11ea-938b-662118190bfe.png)

Reply:
Seems like it is faster, but I implemented it just because I didn't know scipy had its own function.
![Screenshot from 2020-07-17 00-31-40](https://user-images.githubusercontent.com/25507629/87745824-55c9e180-c7c5-11ea-954a-90778b8c5561.png)